src/ast.rs

```diff
@@ -0,0 +1,669 @@
+use super::jsonschema;
+use std::collections::HashMap;
+
+#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum Atom {
+    Boolean,
+    Integer,
+    Number,
+    String,
+    JSON,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Object {
+    pub fields: HashMap<String, Box<Tag>>,
+}
+
+impl Object {
+    pub fn new(fields: HashMap<String, Tag>) -> Self {
+        let boxed: HashMap<String, Box<Tag>> =
+            fields.into_iter().map(|(k, v)| (k, Box::new(v))).collect();
+        Object { fields: boxed }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Array {
+    pub items: Box<Tag>,
```

Comment:
The box is required here because this struct is actually part of a larger recursive type. Without the indirection, the type can be infinitely sized.

```
Tag -> Type::Array(Array { items }) -> Tag
```